### DISCUSSION TITLE
TcpServer和EventLoop该如何正确析构

### DISCUSSION THREAD

**Discussion by Jasons0215** [chenshuo/muduo#597] [General]
你好
问题1：
我看muduo得example里TcpServer和EventLoop都是在主线程里创建并执行，但是在我实际得业务里，有自己得业务循环，所以我将TcpServer和EventLoop放在子线程里去创建和loop，但是程序退出时，会提升和创建得线程不一致导致crash。

问题2：
是否TcpServer得对象需要先于EventLoop对象析构，不然TcpServer里得EventLoop对象就失效了

期待您得回复！！

创建子线程
int Gate::Init(const std::string& conf)
{
    net_thread_ = std::thread(&Gate::StartNetLoop, this);

    signal(SIGTERM, SignalProcess);
}

在子线程：
void Gate::StartNetLoop()
{
    muduo::net::InetAddress serverAddr(tcp_port_);
    event_loop_ = std::make_shared<muduo::net::EventLoop>();
    im_net_server_ = std::make_shared<ImNetServer>(event_loop_.get(), serverAddr, channel_);
    im_net_server_->setThreadNum(net_thread_count_);
    im_net_server_->start();
    event_loop_->loop();
}

在收到kill消息时退出loop
int Gate::Finish()
{   
    event_loop_->runInLoop(std::bind(&muduo::net::EventLoop::quit, event_loop_.get()));
    net_thread_.join();
    return 0;
}

**Comment by chenshuo** [comment]
1. 原则上你可以先在子线程析构 TcpServer 和 EventLoop，再退出。如果你确定 crash 不会对系统造成影响，也可以把相关的检查去掉，让程序安静地退出。
2. 是，EventLoop 的生命期应该长于 TcpServer。
